state/factory/factory_test.go

```diff
@@ -1203,6 +1203,181 @@ func TestDeleteAndPutSameKey(t *testing.T) {
+	priKeyA := identityset.PrivateKey(28)
+
+	tsf1 := action.NewTransfer(big.NewInt(20), b.String(), nil)
+	elp1 := (&action.EnvelopeBuilder{}).SetNonce(1).SetGasLimit(20000).SetAction(tsf1).Build()
+	selp1, err := action.Sign(elp1, priKeyA)
+	require.NoError(err)
+
+	ctx = protocol.WithBlockCtx(
+		ctx,
+		protocol.BlockCtx{
+			BlockHeight: 1,
+			Producer:    identityset.Address(27),
+			GasLimit:    testutil.TestGasLimit,
+		},
+	)
+	ctx = protocol.WithFeatureCtx(ctx)
+	mockActPool := mock_actpool.NewMockActPool(gomock.NewController(t))
+	mockActPool.EXPECT().PendingActionMap().Return(map[string][]*action.SealedEnvelope{
+		a.String(): {selp1},
+	}).Times(1)
+
+	blk1, err := sdb.Mint(
+		protocol.WithBlockchainCtx(
+			ctx,
+			protocol.BlockchainCtx{
+				ChainID: 1,
+				Tip: protocol.TipInfo{
+					Height: 0,
+					Hash:   hash.ZeroHash256,
+				},
+			},
+		),
+		mockActPool,
+		identityset.PrivateKey(27))
+	require.NoError(err)
+	require.NotNil(blk1)
+
+	ws1, exist, err := sdb.(*stateDB).getFromWorkingSets(ctx, blk1.HashBlock())
+	require.NoError(err)
+	require.True(exist)
+	require.NotNil(ws1)
+	// Check balances after the first block
+	accountA, err := accountutil.AccountState(ctx, ws1, a)
+	require.NoError(err)
+	accountB, err := accountutil.AccountState(ctx, ws1, b)
+	require.NoError(err)
+	require.Equal(big.NewInt(80), accountA.Balance)
+	require.Equal(big.NewInt(70), accountB.Balance)
+
+	// Mint the second block with a transfer from B to C
+	c := identityset.Address(30)
+	priKeyB := identityset.PrivateKey(29)
+
+	tsf2 := action.NewTransfer(big.NewInt(30), c.String(), nil)
+	elp2 := (&action.EnvelopeBuilder{}).SetNonce(1).SetGasLimit(20000).SetAction(tsf2).Build()
+	selp2, err := action.Sign(elp2, priKeyB)
+	require.NoError(err)
+
+	mockActPool.EXPECT().PendingActionMap().Return(map[string][]*action.SealedEnvelope{
+		b.String(): {selp2},
+	}).Times(1)
+
+	ctx = protocol.WithBlockCtx(
+		ctx,
+		protocol.BlockCtx{
+			BlockHeight: 2,
+			Producer:    identityset.Address(27),
+			GasLimit:    testutil.TestGasLimit,
+		},
+	)
+	ctx = protocol.WithFeatureCtx(ctx)
+	blk2, err := sdb.Mint(
+		protocol.WithBlockchainCtx(
+			ctx,
+			protocol.BlockchainCtx{
+				ChainID: 1,
+				Tip: protocol.TipInfo{
+					Height: 1,
+					Hash:   blk1.HashBlock(),
+				},
+			},
+		),
+		mockActPool,
+		identityset.PrivateKey(27),
+	)
+	require.NoError(err)
+	require.NotNil(blk2)
+
+	ws2, exist, err := sdb.(*stateDB).getFromWorkingSets(ctx, blk2.HashBlock())
+	require.NoError(err)
+	require.True(exist)
+	require.NotNil(ws2)
+
+	// Check balances after the second block
+	accountA, err = accountutil.AccountState(ctx, ws2, a)
+	require.NoError(err)
+	accountB, err = accountutil.AccountState(ctx, ws2, b)
+	require.NoError(err)
+	accountC, err := accountutil.AccountState(ctx, ws2, c)
+	require.NoError(err)
+	require.Equal(big.NewInt(80), accountA.Balance)
+	require.Equal(big.NewInt(40), accountB.Balance)
+	require.Equal(big.NewInt(30), accountC.Balance)
```

Comment:
put these into a common func `checkBalanaceABC(ws, expectBalance [3]big.Int)`? then call it after diff actions
```
checkBalanaceABC(ws2, [3]big.Int{80, 40, 30})
checkBalanaceABC(ws1, [3]big.Int{80, 70, 0})
sdb.PutBlock(ctx, blk1)
checkBalanaceABC(sdb, [3]big.Int{80, 70, 0})
... and more
```